designs/2021-fixable-disable-directives/design.md

```diff
@@ -0,0 +1,136 @@
+  - These problems have a `ruleId` of `null`
+- `_verifyWithoutProcessors` is called within the call stack of each of the 1-10 passes in the linter's `verifyAndFix`
+
+This RFC proposes making three changes:
+
+- In the patched `options.fix`, consider a problem's meta type to be `"directive"` if its `ruleId` is `null`
+- Pass the `SourceCode` being linted as a parameter to `applyDisableDirectives`
+- Add a `fix` method to the unused directive problems returned by `applyDirectives` that uses the `SourceCode`
+
+### Fix Behavior
+
+Directives where at least one rule is still used will have only the unused rule names removed from their source text.
+
+Directives where all >=1 rules are unused will use the `SourceCode` to compute:
+
+- If they are the only non-whitespace on their line, delete that line
+- Otherwise, delete just the comment and any now-unnecessary surrounding whitespace
+
+#### Fix Behavior Examples
+
+```diff
+- /* eslint-disable */
+```
+
+```diff
+- // eslint-disable-next-line -- related explanation
+```
+
+```diff
+- // eslint-disable-next-line unused, used
++ // eslint-disable-next-line used
+```
+
+```diff
+- before /* eslint-disable-next-line -- related explanation */ after
++ before after
+```
+
+```diff
+- before // eslint-disable-next-line -- related explanation
++ before
+```
+
+```diff
+  // before
+- // eslint-disable-next-line
+  // after
+```
+
+Multiline block comments are already not allowed to be disable directives. [eslint/eslint#10334](https://github.com/eslint/eslint/issues/10334)
+
+## Documentation
+
+- [Command Line Interface > Fixing Problems](https://eslint.org/docs/user-guide/command-line-interface#fixing-problems)'s `--fix` documentation should mention the added directives fixing and the new `--fix-type`
+- [Rules > Report Unused ESLint Disable Comments](https://eslint.org/docs/user-guide/configuring/rules#report-unused-eslint-disable-comments) should link to that documentation
+
+## Drawbacks
+
+Like any new feature, this flag will slightly increase the complexity and maintenance costs of ESLint core.
+
+This RFC's implementation would lock in the name for a new `--fix-type` even though we only have one concrete use case for it.
+
+## Backwards Compatibility Analysis
+
+It is unlikely but not impossible that some `--fix` dependant users would rely on unused disable directives remaining in code.
+
+Otherwise, this change is additive in behavior.
+
+## Alternatives
+
+- Applying directive fixes after the `verifyAndFix` passes
+  - Not ideal: deleting a comment could introduce new rule violations that would warrant running >=1 other pass
+- Writing an external tool to apply these options
+  - Not ideal: the internal implementation is simple; external would have to deal with variant formatters, etc and suffer from the same introduced rule violations
+
+## Open Questions
+
+- Is `"directive"` a good name for the fix type?
```

Comment:
The only other name I'm thinking of is `"eslint-disable"`, which is even more specific to this particular use if that's desirable.